design/secrets.md

```diff
@@ -0,0 +1,81 @@
+# Design proposal template (replace with your proposal's title)
+
+Currently, Velero only supports a single AIM access credential secret per location provider/plugin. Velero creates and stores the plugin credential secret under the hard-coded key `secret.cloud-credentials.data.cloud`.
+
+This makes it so switching from one plugin to another necessitates overriding the existing credential secret with the appropriate one for the new plugin.
+
+## Goals
+
+- To allow Velero to create and store multiple secrets for provider credentials, even multiple credentials for the same provider
+- To improve the UX for configuring the velero deployment with multiple plugins/providers, and corresponding IAM secrets.
```

Comment:
This line mentions IAM secrets, though we don't discuss them anywhere else in the document.

What can we do to make that experience better? Often, IAM credentials are applied at the node level and may not be Kubernetes secrets.

Based on https://github.com/vmware-tanzu/velero-plugin-for-aws#option-2-set-permissions-using-kube2iam, it looks like that would be up to kube2iam or similar. It may also be the case that IAM scenarios are not covered by this particular design.

Reply:
The lack of discussion on IAM credentials is my fault. The initial set of goals was written by @carlisia and I didn't expand on that topic. Now that I look at the link you posted, I don't think the kube2iam approach is compatible with the case of having multiple credentials for AWS managed by Velero. An IAM role could be created which supports multiple sets of credentials, but that isn't something that is controllable through Velero. And I'm not sure if we would want to introduce such specific AWS functionality into Velero.

Reply:
I would rather have the IAM case be handled separately from this design change if possible.

Reply:
I think I wrote that down.

Reply:
Yeah, I agree on not introducing any specific infra provider functionality into it; I was using AWS as a popular example. However, [GCP has a similar feature](https://github.com/vmware-tanzu/velero-plugin-for-gcp#option-2-set-permissions-with-using-workload-identity-optional), and [so does Azure](https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure#option-2-use-aad-pod-identity).

They're called different things, but I think the idea is the same in that they all three tie the auth to the nodes/pods instead of via a credentials file. A brief skim of the READMEs says that the Azure one is the only one that has K8s resources associated directly in this flow, but the others may have some behind the scenes that aren't covered in the readme.

Reply:
> I would rather have the IAM case be handled separately from this design change if possible.

100% fine with that. I think we can call that out explicitly in the non-goals section. Something like "Node-based auth features" or similar so that we're not just focused on the AWS solution.

Reply:
Hmm, the Azure one looks similar to the existing workflow, the outcome of using AAD pod identity is to create a credentials file which is then used with `velero` install. It's not clear to me if multiple roles/identities can be handled through one file in that method.

The others though are managed outside of velero though, and I think would continue to work as before. If you had a BSL that used a plugin which included those methods of specifying credentials, you would say that no credentials are needed and then no credentials would be passed through in the environment. I think if you needed to access other accounts using those plugins, it might be possible to add other credentials, and use them in the environment for the plugin process (overriding the credentials for the pod/deployment).